src/crypto/tls/handshake_client_tls13.go

```diff
@@ -245,6 +284,45 @@ func (hs *clientHandshakeStateTLS13) processHelloRetryRequest() error {
 		hs.transcriptInner.Reset()
 		hs.transcriptInner.Write([]byte{typeMessageHash, 0, 0, uint8(len(chHash))})
 		hs.transcriptInner.Write(chHash)
+
+		// Check for ECH acceptance confirmation.
+		if len(hs.serverHello.ech) == 8 {
+			echAcceptConfHRRTranscript := cloneHash(hs.transcriptInner, hs.suite.hash)
+			if echAcceptConfHRRTranscript == nil {
+				c.sendAlert(alertInternalError)
+				return errors.New("tls: internal error: failed to clone hash")
+			}
+
+			signal := hs.serverHello.ech
+			hs.serverHello.ech = []byte{0, 0, 0, 0, 0, 0, 0, 0}
+			hs.serverHello.raw = nil
+			echAcceptConfHRRTranscript.Write(hs.serverHello.marshal())
```

Comment:
Will this marshal call ensure that the order in which extensions was received from the server is maintained? (Surely it must since the tests are presumably passing, but I thought I'd ask just in case.)

Reply:
Good catch! Added a new function, echEncodeAcceptConfHelloRetryRequest(), that encodes this properly.

No, this isn't guaranteed, and the tests don't catch this. I've been meaning to think about how design tests for this but haven't gotten around to it.